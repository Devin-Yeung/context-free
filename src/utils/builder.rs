use crate::utils::symbols;
use bnf::{Grammar, Term};
use once_cell::sync::OnceCell;
use std::collections::{HashMap, HashSet};

pub fn epsilon() -> &'static Term {
    static EPSILON: OnceCell<Term> = OnceCell::new();
    EPSILON.get_or_init(|| Term::Terminal(String::from("ε")))
}

fn dollar() -> &'static Term {
    static DOLLAR: OnceCell<Term> = OnceCell::new();
    DOLLAR.get_or_init(|| Term::Terminal(String::from("$")))
}

pub struct FirstFollowBuilder<'grammar> {
    pub(crate) inner: HashMap<&'grammar Term, HashSet<&'grammar Term>>,
}

impl<'grammar> FirstFollowBuilder<'grammar> {
    pub(crate) fn new(grammar: &'grammar Grammar) -> FirstFollowBuilder<'grammar> {
        let mut inner = HashMap::new();

        // initialize the table
        symbols(&grammar)
            .into_iter()
            .filter(|term| term != &epsilon()) // epsilon is a special non-terminal
            .for_each(|term| {
                inner.insert(term, HashSet::new());
            });

        FirstFollowBuilder { inner }
    }

    // Insert term to First(x)
    ///
    /// return true if the First(x) changes
    /// otherwise return false
    pub(crate) fn insert_term(&mut self, x: &'grammar Term, term: &'grammar Term) -> bool {
        // First(x)
        let first_x = self.inner.get_mut(x).unwrap();

        // Insert term to First(x)
        let before = first_x.len();
        first_x.insert(term);
        let after = first_x.len();

        // check if set changes
        before != after
    }

    // Insert epsilon to First(x)
    pub(crate) fn insert_epsilon(&mut self, x: &'grammar Term) -> bool {
        self.insert_term(x, epsilon())
    }

    // Insert dollar to Follow(x)
    pub(crate) fn insert_dollar(&mut self, x: &'grammar Term) -> bool {
        self.insert_term(x, dollar())
    }

    /// First(x)
    pub(crate) fn first(&self, x: &Term) -> HashSet<&'grammar Term> {
        self.inner
            .get(x)
            .map_or_else(|| HashSet::new(), |set| set.clone())
    }

    /// Follow(x)
    pub(crate) fn follow(&self, x: &Term) -> HashSet<&'grammar Term> {
        self.inner
            .get(x)
            .map_or_else(|| HashSet::new(), |set| set.clone())
    }

    pub(crate) fn insert_set(&mut self, x: &'grammar Term, set: HashSet<&'grammar Term>) -> bool {
        // First(x)
        let first_x = self.inner.get_mut(x).unwrap();

        // Insert set into First(x)
        let before = first_x.len();
        first_x.extend(set);
        let after = first_x.len();

        // check if set changes
        return before != after;
    }

    /// Insert First(y) \ { ε } into First(x)
    ///
    /// return true if the First(x) changes
    /// otherwise return false
    pub(crate) fn insert_first_no_epsilon(&mut self, x: &'grammar Term, y: &'grammar Term) -> bool {
        // First(y)
        let mut first_y = self.first(y);
        // First(y) \ { ε }
        first_y.remove(epsilon());
        // Insert First(y) \ { ε } into First(x)
        self.insert_set(x, first_y)
    }

    /// Insert Follow(y) into Follow(x)
    pub(crate) fn insert_follow(&mut self, x: &'grammar Term, y: &'grammar Term) -> bool {
        // Follow(y)
        let follow_y = self.follow(y);
        println!("Insert {:?} to Follow({})", follow_y, x);
        // Insert Follow(y) into Follow(x)
        self.insert_set(x, follow_y)
    }

    pub(crate) fn build(self) -> HashMap<&'grammar Term, HashSet<&'grammar Term>> {
        self.inner
    }
}
