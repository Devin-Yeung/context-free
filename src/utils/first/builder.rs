use crate::utils::first::First;
use crate::utils::{epsilon, symbols};
use bnf::{Grammar, Production, Term};
use std::cell::RefCell;
use std::collections::{HashMap, HashSet};

pub struct FirstBuilder<'grammar> {
    pub(crate) grammar: &'grammar Grammar,
    pub(crate) first: RefCell<HashMap<&'grammar Term, HashSet<&'grammar Term>>>,
    pub(crate) lookup: HashMap<&'grammar Term, &'grammar Production>,
}

impl<'grammar> FirstBuilder<'grammar> {
    pub(crate) fn new(grammar: &'grammar Grammar) -> FirstBuilder<'grammar> {
        let mut first = HashMap::new();

        let lookup = grammar
            .productions_iter()
            .map(|production| (&production.lhs, production))
            .collect::<HashMap<_, _>>();

        // initialize the table
        symbols(grammar)
            .filter(|term| term != &epsilon()) // epsilon is a special non-terminal
            .for_each(|term| {
                first.insert(term, HashSet::new());
            });

        let first = RefCell::new(first);

        FirstBuilder {
            grammar,
            first,
            lookup,
        }
    }

    fn build_first(&mut self) {
        symbols(self.grammar)
            .filter(|term| term != &epsilon())
            .for_each(|t| {
                match t {
                    Term::Terminal(s) => {
                        // Rule1: If X is a terminal, then First(X) = { X }
                        self.insert_term(t, t);
                        println!("Rule1: Push {} to First({})", s, t);
                    }
                    Term::Nonterminal(_) => { /* skip */ }
                };

                if self.produce_epsilon(t) {
                    // Rule2: If X is an ε-production, then add ε to First(X)
                    self.insert_epsilon(t);
                    println!("Rule2: Push ε to First({})", t);
                }
            });

        loop {
            let mut changed = false;

            symbols(self.grammar)
                .filter(|term| matches!(*term, Term::Nonterminal(_)))
                .for_each(|lhs| {
                    println!("===> Checking Symbol: {}", lhs);
                    let production = self.lookup.get(lhs).unwrap();
                    // Rule3: If X is a non-terminal and X → Y1 Y2 ... Yk,
                    // then add First(Y1) ∖ {ε} to First(X)
                    for expr in production.rhs_iter() {
                        for term in expr
                            .terms_iter()
                            .filter(|term| term != &&Term::Terminal("ε".to_string()))
                        {
                            // First(Y1) ∖ {ε} to First(X)
                            changed |= self.insert_first_no_epsilon(&production.lhs, term);
                            println!(
                                "Rule3/4: Push First({}) \\ ε to First({})",
                                term, production.lhs
                            );
                            // terminate (check next expression) if X does NOT produce ε
                            if !self.produce_epsilon(term) {
                                println!("{} does NOT produce ε", term);
                                break;
                            }
                        }
                        // Rule 5: If X is a non-terminal and X -> Y1 Y2 ... Yk,
                        // and First(Yi) produce ε for all i, then add ε to First(X)
                        if expr.terms_iter().all(|term| self.produce_epsilon(term)) {
                            println!("Rule5: Push ε to First({})", production.lhs);
                            changed = self.insert_epsilon(&production.lhs);
                        }
                    }
                });

            if !changed {
                println!("Unchanged, break!");
                break;
            }
        } // End of loop
    }

    fn produce_epsilon(&self, term: &Term) -> bool {
        let production = self.lookup.get(&term);
        if production.is_none() {
            return false;
        }

        let production = production.unwrap();

        match &production.lhs {
            Term::Terminal(t) => {
                if t == "ε" {
                    return true;
                }
            }
            Term::Nonterminal(nt) => {
                if nt == "ε" {
                    return true;
                }
            }
        }

        production
            .rhs_iter()
            .map(|expr| {
                expr.terms_iter().all(|term| match term {
                    Term::Terminal(t) => t == "ε",
                    Term::Nonterminal(nt) => nt == "ε",
                })
            })
            .any(|v| v)
    }

    // Insert term to First(x)
    ///
    /// return true if the First(x) changes
    /// otherwise return false
    fn insert_term(&self, x: &'grammar Term, term: &'grammar Term) -> bool {
        let mut first = self.first.borrow_mut();
        // First(x)
        dbg!(x);
        let first_x = first.get_mut(x).unwrap();

        // Insert term to First(x)
        let before = first_x.len();
        first_x.insert(term);
        let after = first_x.len();

        // check if set changes
        before != after
    }

    // Insert epsilon to First(x)
    fn insert_epsilon(&self, x: &'grammar Term) -> bool {
        self.insert_term(x, epsilon())
    }

    /// First(x)
    fn first(&self, x: &Term) -> HashSet<&'grammar Term> {
        self.first
            .borrow()
            .get(x)
            .map_or_else(HashSet::new, |set| set.clone())
    }

    fn insert_set(&self, x: &'grammar Term, set: HashSet<&'grammar Term>) -> bool {
        let mut first = self.first.borrow_mut();
        // First(x)
        let first_x = first.get_mut(x).unwrap();

        // Insert set into First(x)
        let before = first_x.len();
        first_x.extend(set);
        let after = first_x.len();

        // check if set changes
        before != after
    }

    /// Insert First(y) \ { ε } into First(x)
    ///
    /// return true if the First(x) changes
    /// otherwise return false
    fn insert_first_no_epsilon(&self, x: &'grammar Term, y: &'grammar Term) -> bool {
        // First(y)
        let mut first_y = self.first(y);
        // First(y) \ { ε }
        first_y.remove(epsilon());
        // Insert First(y) \ { ε } into First(x)
        self.insert_set(x, first_y)
    }

    pub(crate) fn build(mut self) -> First<'grammar> {
        self.build_first();
        First {
            first: self.first.into_inner(),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::utils::first::builder::FirstBuilder;
    use bnf::Term;

    #[test]
    fn first() {
        let grammar = r#"
        <E> ::= <T> <E'>
        <E'> ::= '+' <T> <E'> | 'ε'
        <T> ::= <F> <T'>
        <T'> ::= '*' <F> <T'> | 'ε'
        <F> ::= '(' <E> ')' | 'id'
        "#
        .parse()
        .unwrap();

        let first = FirstBuilder::new(&grammar).build();
        first.first.iter().for_each(|(lhs, rhs)| match lhs {
            Term::Terminal(_) => {
                assert_eq!(rhs.len(), 1)
            }
            Term::Nonterminal(_) => {
                assert_eq!(rhs.len(), 2)
            }
        })
    }
}
