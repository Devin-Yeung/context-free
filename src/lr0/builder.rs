use crate::lr0::core::{LR0Closure, LR0Item, LR0ItemSet};
use crate::utils::symbols;
use bnf::{Grammar, Production, Term};
use std::collections::{HashMap, VecDeque};

pub struct LR0Builder<'grammar> {
    grammar: &'grammar Grammar,
    closures: Vec<LR0ItemSet<'grammar>>,
    transitions: HashMap<(usize, &'grammar Term), usize>,
}

impl<'grammar> LR0Builder<'grammar> {
    pub fn new(grammar: &'grammar Grammar) -> LR0Builder<'grammar> {
        LR0Builder {
            grammar,
            closures: Vec::new(),
            transitions: HashMap::new(),
        }
    }

    pub fn build(mut self, augmentation: &'grammar Production) -> LR0Closure {
        let initial = LR0ItemSet::from_iter(vec![LR0Item::from_production(augmentation).unwrap()]);
        self.build_closure(&initial);
        self.build_transition();
        LR0Closure {
            closures: self.closures,
            transitions: self.transitions,
        }
    }

    fn build_closure(&mut self, initial: &LR0ItemSet<'grammar>) {
        let mut waiting = VecDeque::from([initial.closure(self.grammar)]);
        self.closures.push(initial.closure(self.grammar));

        while !waiting.is_empty() {
            let cur = waiting.pop_front().unwrap();
            symbols(self.grammar).iter().for_each(|term| {
                let goto = cur.goto(self.grammar, term);
                if !goto.items.is_empty() && !self.contains(&goto) {
                    self.closures.push(goto);
                    waiting.push_back(self.closures.last().unwrap().clone());
                }
            });
        }
    }

    fn build_transition(&mut self) {
        for closure in &self.closures {
            for term in symbols(self.grammar) {
                let goto = closure.goto(self.grammar, term);
                if goto.items.is_empty() {
                    continue;
                }
                let goto_index = self.index_of(&goto);
                let cur_index = self.index_of(closure);
                self.transitions.insert((cur_index, term), goto_index);
                println!("goto(I_{}, {}) = I_{}", cur_index, term, goto_index);
            }
        }
    }

    fn contains(&self, set: &LR0ItemSet<'grammar>) -> bool {
        self.closures.iter().any(|v| v == set)
    }

    fn index_of(&self, set: &LR0ItemSet<'grammar>) -> usize {
        self.closures.iter().position(|v| v == set).unwrap()
    }
}

#[cfg(test)]
mod tests {
    use crate::lr0::core::LR0Closure;
    use bnf::Production;
    use std::str::FromStr;

    #[test]
    fn it_works() {
        let grammar = r#"
        <E'> ::= <E>
        <E> ::= <E> '+' <T> | <T>
        <T> ::= <T> '*' <F> | <F>
        <F> ::= '(' <E> ')' | 'id'
        "#
        .parse()
        .unwrap();
        let augmentation = Production::from_str("<E'> ::= <E>").unwrap();

        assert_eq!(
            LR0Closure::new(&grammar, &augmentation).closures().len(),
            12
        );
    }

    #[test]
    fn object_clause() {
        let grammar = r#"
        <S> ::= <O> 'v' <C>
        <O> ::= 'n'
        <C> ::= <S>
        <C> ::= 'n'
        "#
        .parse()
        .unwrap();
        let augmentation = Production::from_str("<S'> ::= <S>").unwrap();

        assert_eq!(LR0Closure::new(&grammar, &augmentation).closures().len(), 8);
    }
}
