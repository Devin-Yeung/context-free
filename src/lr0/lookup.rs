use bnf::{Expression, Grammar, Term};

type LR0Production<'grammar> = (&'grammar Term, &'grammar Expression);

#[derive(Debug)]
pub struct Lookup<'grammar> {
    grammar: &'grammar Grammar,
    // lookup: HashMap<&'grammar Term, &'grammar Production>,
}

impl<'grammar> Lookup<'grammar> {
    pub fn new(grammar: &'grammar Grammar) -> Lookup<'grammar> {
        Lookup { grammar }
    }

    pub fn get(&self, term: &Term) -> impl IntoIterator<Item = LR0Production<'grammar>> {
        let prod = self
            .grammar
            .productions_iter()
            .filter(|production| production.lhs == *term)
            .flat_map(|production| production.rhs_iter().map(|expr| (&production.lhs, expr)))
            .collect::<Vec<_>>();
        prod
    }

    pub fn productions(&self) -> impl IntoIterator<Item = LR0Production<'grammar>> {
        let prod = self
            .grammar
            .productions_iter()
            .flat_map(|production| production.rhs_iter().map(|expr| (&production.lhs, expr)))
            .collect::<Vec<_>>();
        prod
    }
}

#[cfg(test)]
mod tests {
    use crate::lr0::lookup::{LR0Production, Lookup};
    use bnf::{Grammar, Term};

    pub fn grammar() -> Grammar {
        let input = r#"
        <E'> ::= <E>
        <E> ::= <E> '+' <T> | <T>
        <T> ::= <T> '*' <F> | <F>
        <F> ::= '(' <E> ')' | 'id'
        "#;
        let grammar: Grammar = input.parse().unwrap();
        grammar
    }

    #[test]
    fn productions() {
        let grammar = grammar();
        let lookup = Lookup::new(&grammar);
        let production: Vec<LR0Production> = lookup.productions().into_iter().collect::<Vec<_>>();
        assert_eq!(production.len(), 7);
    }

    #[test]
    fn term_productions() {
        let grammar = grammar();
        let lookup = Lookup::new(&grammar);
        let production: Vec<LR0Production> = lookup
            .get(&Term::Nonterminal("E".to_string()))
            .into_iter()
            .collect::<Vec<_>>();
        assert_eq!(production.len(), 2);
    }
}
