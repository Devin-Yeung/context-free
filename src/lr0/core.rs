use crate::lr0::builder::LR0Builder;
use crate::lr0::lookup::Lookup;
use bnf::{Expression, Grammar, Production, Term};
use itertools::Itertools;
use std::collections::{HashMap, HashSet};
use std::fmt::{Display, Formatter};
use std::hash::Hash;

use tabled::builder::Builder;
use tabled::Table;

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
pub struct LR0Item<'grammar> {
    pub(crate) lhs: &'grammar Term,
    pub(crate) rhs: &'grammar Expression,
    pub(crate) delimiter: usize,
}

pub struct LR0Closure<'grammar> {
    pub(crate) closures: Vec<LR0ItemSet<'grammar>>,
    pub(crate) transitions: HashMap<(usize, &'grammar Term), usize>,
}

impl<'grammar> LR0Closure<'grammar> {
    pub fn new(
        grammar: &'grammar Grammar,
        augmentation: &'grammar Production,
    ) -> LR0Closure<'grammar> {
        LR0Builder::new(grammar).build(augmentation)
    }

    pub fn closures(&self) -> &Vec<LR0ItemSet<'grammar>> {
        &self.closures
    }

    pub fn closure_table(&self) -> Table {
        let mut builder = Builder::default();
        for set in self.closures.iter() {
            builder.push_record([set.to_string()]);
        }
        builder.index().build()
    }

    /// enumerate all the `from` states
    /// which means `(from, term) = to`
    pub fn enumerate_states(&self) -> impl Iterator<Item = usize> {
        self.transitions
            .keys()
            .map(|(i, _)| *i)
            .chain(self.transitions.values().map(|i| *i))
            .unique()
            .sorted()
    }

    pub fn transition_table(&self) -> Table {
        let mut builder = Builder::default();
        let header = self
            .transitions
            .keys()
            .map(|(_, t)| t)
            .unique()
            .sorted()
            .collect::<Vec<_>>();
        // header
        builder.set_header(
            std::iter::once(String::from("Closure/Symbol"))
                .chain(header.iter().map(|t| t.to_string())),
        );

        self.enumerate_states().for_each(|from| {
            let row = std::iter::once(format!("I_{}", from))
                .chain(header.iter().map(|term| {
                    match self.transition(from, term) {
                        None => String::new(), /* ∅ */
                        Some(to) => format!("I_{}", to),
                    }
                }))
                .collect::<Vec<_>>();
            builder.push_record(row);
        });

        builder.build()
    }

    pub fn enumerate_lr0<'a>(&'a self) -> impl Iterator<Item = (usize, &'a LR0Item<'grammar>)> {
        self.closures
            .iter()
            .enumerate()
            .flat_map(|(i, set)| set.items.iter().map(move |item| (i, item)))
    }

    /// number of states in the closure
    pub fn len(&self) -> usize {
        self.closures.len()
    }

    pub fn transition<'a>(&self, from: usize, via: &Term) -> Option<usize> {
        self.transitions.get(&(from, via)).copied()
    }
}

impl<'grammar> LR0Item<'grammar> {
    pub fn from_production(production: &'grammar Production) -> Option<LR0Item<'grammar>> {
        if production.rhs_iter().count() != 1 {
            return None;
        }
        let rhs = production.rhs_iter().next().unwrap();
        Some(LR0Item {
            lhs: &production.lhs,
            rhs,
            delimiter: 0,
        })
    }
}

impl<'grammar> Display for LR0Item<'grammar> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let mut s = Vec::new();
        s.push(self.lhs.to_string());
        s.push("->".to_string());
        let mut rhs = self
            .rhs
            .terms_iter()
            .map(|t| t.to_string())
            .collect::<Vec<_>>();
        rhs.insert(self.delimiter, "•".to_string());
        s.extend(rhs);
        f.write_str(&s.join(" "))
    }
}

impl<'grammar> Display for LR0ItemSet<'grammar> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        let s = self
            .items
            .iter()
            .map(|i| i.to_string())
            .sorted() // make display result deterministic
            .collect::<Vec<_>>()
            .join(", ");
        f.write_fmt(format_args!("[{}]", s))
    }
}

impl<'grammar> LR0Item<'grammar> {
    pub fn expect(&self) -> Option<&'grammar Term> {
        self.rhs.terms_iter().nth(self.delimiter)
    }
}

#[derive(Debug, Clone, Eq, PartialEq)]
pub struct LR0ItemSet<'grammar> {
    pub(crate) items: HashSet<LR0Item<'grammar>>,
}

impl<'grammar> FromIterator<LR0Item<'grammar>> for LR0ItemSet<'grammar> {
    fn from_iter<T: IntoIterator<Item = LR0Item<'grammar>>>(iter: T) -> Self {
        Self {
            items: iter.into_iter().collect::<HashSet<_>>(),
        }
    }
}

impl<'grammar> LR0ItemSet<'grammar> {
    pub fn new() -> Self {
        Self {
            items: HashSet::new(),
        }
    }
    pub fn closure(&self, grammar: &'grammar Grammar) -> LR0ItemSet<'grammar> {
        let lookup = Lookup::new(grammar);

        let mut closure = self.clone();

        loop {
            let mut extend = HashSet::new();

            for item in &closure.items {
                if let Some(x) = item.rhs.terms_iter().nth(item.delimiter) {
                    // x is the term after dot
                    for production in lookup.get(x) {
                        let lr0_item = LR0Item {
                            lhs: production.0,
                            rhs: production.1,
                            delimiter: 0,
                        };
                        if !closure.contains(&lr0_item) {
                            extend.insert(lr0_item);
                        }
                    }
                }
            }

            // check if closure change or not
            if extend.is_empty() {
                break;
            } else {
                closure.items.extend(extend);
            }
        }

        closure
    }

    pub fn goto(&self, grammar: &'grammar Grammar, term: &Term) -> LR0ItemSet<'grammar> {
        let items = self
            .items
            .iter()
            .filter_map(|item| {
                if item.expect() == Some(term) {
                    let mut bump = item.clone();
                    bump.delimiter += 1;
                    Some(bump)
                } else {
                    None
                }
            })
            .collect::<HashSet<_>>();
        let set = LR0ItemSet { items };
        set.closure(grammar)
    }

    pub fn contains(&self, item: &LR0Item<'grammar>) -> bool {
        self.items.contains(item)
    }
}

#[cfg(test)]
mod test {
    use crate::lr0::core::{LR0Closure, LR0Item, LR0ItemSet};
    use bnf::{Expression, Grammar, Production, Term};
    use std::str::FromStr;

    pub fn grammar() -> Grammar {
        let input = r#"
        <E'> ::= <E>
        <E> ::= <E> '+' <T> | <T>
        <T> ::= <T> '*' <F> | <F>
        <F> ::= '(' <E> ')' | 'id'
        "#;
        let grammar: Grammar = input.parse().unwrap();
        grammar
    }

    #[test]
    fn it_works() {
        let grammar = grammar();
        let augmentation = Production::from_str("<E'> ::= <E>").unwrap();
        let closure = LR0Closure::new(&grammar, &augmentation).closure_table();
        insta::assert_display_snapshot!(closure);
    }

    #[test]
    fn deterministic_transition_table() {
        let grammar = grammar();
        let augmentation = Production::from_str("<E'> ::= <E>").unwrap();
        let transitions = LR0Closure::new(&grammar, &augmentation).transition_table();
        insta::assert_display_snapshot!(transitions);
    }

    #[test]
    fn more_items() {
        let set = [("<E'>", "<E>", 1usize), ("<E>", "<E> '+' <T>", 1usize)];
        let set = set
            .into_iter()
            .map(|(lhs, rhs, delimiter)| {
                let lhs = Term::from_str(lhs).unwrap();
                let rhs = Expression::from_str(rhs).unwrap();
                (lhs, rhs, delimiter)
            })
            .collect::<Vec<_>>();

        let lr0_set: LR0ItemSet =
            LR0ItemSet::from_iter(set.iter().map(|(lhs, rhs, delimiter)| LR0Item {
                lhs,
                rhs,
                delimiter: *delimiter,
            }));
        assert_eq!(lr0_set.closure(&grammar()).items.len(), 2);
    }

    #[test]
    fn goto() {
        let grammar = grammar();
        let lhs = Term::from_str("<E'>").unwrap();
        let rhs = Expression::from_str("<E>").unwrap();

        let lr0_item = LR0Item {
            lhs: &lhs,
            rhs: &rhs,
            delimiter: 0,
        };

        let set = LR0ItemSet::from_iter(vec![lr0_item]);
        let I_0 = set.closure(&grammar);
        [
            ("<E>", 2usize),
            ("<T>", 2),
            ("<F>", 1),
            ("'('", 7),
            ("'id'", 1),
            ("<E'>", 0),
            ("'+'", 0),
            ("'*'", 0),
            ("')'", 0),
        ]
        .iter()
        .for_each(|(t, cnt)| {
            let term = Term::from_str(t).unwrap();
            let goto = I_0.goto(&grammar, &term);
            assert_eq!(goto.items.len(), *cnt)
        });
    }
}
