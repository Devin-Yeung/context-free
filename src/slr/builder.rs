use crate::lr0::core::{LR0Closure, LR0Item};
use crate::slr::core::SLRInstruction;
use crate::slr::helper::IndexedGrammar;
use crate::utils::dollar;
use crate::utils::follow::Follow;
use bnf::{Grammar, Production, Term};
use std::cell::RefCell;
use std::collections::HashMap;
use std::fmt::{format, Display, Formatter};
use std::iter::{once, repeat};
use tabled::builder::Builder;
use tabled::Table;

pub struct SLRTable<'grammar> {
    grammar: IndexedGrammar<'grammar>,
    table: Vec<HashMap<&'grammar Term, SLRInstruction>>,
}

impl<'grammar> SLRTable<'grammar> {
    pub fn grammar_table(&self) -> Table {
        self.grammar.grammar_table()
    }

    pub fn parsing_table(&self) -> Table {
        let mut builder = Builder::default();

        let header = self
            .grammar
            .terminals()
            .chain(once(dollar()))
            .chain(self.grammar.non_terminals())
            .collect::<Vec<_>>();

        builder.set_header(header.iter().map(|t| t.to_string()));

        self.table.iter().enumerate().for_each(|(_, table)| {
            let row = header
                .iter()
                .map(|t| table.get(t).unwrap_or(&SLRInstruction::Empty))
                .collect::<Vec<_>>();
            builder.push_record(row);
        });

        builder.index().build()
    }
}

impl<'grammar> Display for SLRTable<'grammar> {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        f.write_fmt(format_args!("Grammar: \n{}\n", self.grammar_table()))?;
        f.write_fmt(format_args!("Table: \n{}", self.parsing_table()))
    }
}

pub struct SLRTableBuilder<'grammar> {
    grammar: IndexedGrammar<'grammar>,
    follow: Follow<'grammar>,
    closure: LR0Closure<'grammar>,
    table: RefCell<Vec<HashMap<&'grammar Term, SLRInstruction>>>,
}

impl<'grammar> SLRTableBuilder<'grammar> {
    pub fn new(
        grammar: &'grammar Grammar,
        augmentation: &'grammar Production,
    ) -> SLRTableBuilder<'grammar> {
        let follow = Follow::new(grammar, &augmentation.lhs);
        let closure = LR0Closure::new(grammar, augmentation);
        let grammar = IndexedGrammar::new(grammar);
        let table = RefCell::new(
            repeat(HashMap::<&Term, SLRInstruction>::new())
                .take(closure.len())
                .collect(),
        );
        SLRTableBuilder {
            grammar,
            follow,
            closure,
            table,
        }
    }

    fn shift(&self, from: usize, via: &'grammar Term) {
        let to = self.closure.transition(from, via).unwrap(); // TODO: really not sure whether this unwrap is safe?
        println!("Shift: goto(I_{}, {}) = I_{}", from, via, to);
        let mut table = self.table.borrow_mut();
        table
            .get_mut(from)
            .unwrap()
            .insert(via, SLRInstruction::Shift(to));
    }

    fn reduce(&self, index: usize, lr0: &LR0Item) {
        let grammar_index = self.grammar.get_index_of(&lr0.rhs).unwrap();
        let prod = self.grammar.get(lr0.rhs).unwrap();
        let mut table = self.table.borrow_mut();
        for term in self.follow.follow_of(&prod.lhs).collect::<Vec<_>>() {
            table[index].insert(term, SLRInstruction::Reduce(grammar_index));
            println!(
                "Reduce: set (I_{}, {}) = r{}",
                index,
                term.to_string(),
                grammar_index
            );
        }
    }

    fn goto(&self, from: usize, via: &'grammar Term) {
        debug_assert!(matches!(via, Term::Nonterminal(_)));
        let to = self.closure.transition(from, via).unwrap(); // TODO: really not sure whether this unwrap is safe?
        println!("Goto: goto(I_{}, {}) = I_{}", from, via, to);
        let mut table = self.table.borrow_mut();
        table
            .get_mut(from)
            .unwrap()
            .insert(via, SLRInstruction::Goto(to));
    }

    pub fn build(self) -> SLRTable<'grammar> {
        self.closure.enumerate_lr0().for_each(|(i, lr0)| {
            // take a look at the char after the •
            match lr0.expect() {
                // lr0 expect non character, which is the form of A -> 𝛼 •
                None => {
                    /* Reduce */
                    self.reduce(i, lr0)
                }
                Some(t) => {
                    match t {
                        Term::Terminal(_) => self.shift(i, t),   /* Shift */
                        Term::Nonterminal(_) => self.goto(i, t), /* Goto */
                    }
                }
            }
        });
        SLRTable {
            grammar: self.grammar,
            table: self.table.into_inner(),
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::slr::builder::SLRTableBuilder;
    use bnf::Production;
    use std::str::FromStr;

    #[test]
    fn it_works() {
        let grammar = r#"
        <E'> ::= <E>
        <E> ::= <E> '+' <T> | <T>
        <T> ::= <T> '*' <F> | <F>
        <F> ::= '(' <E> ')' | 'id'
        "#
        .parse()
        .unwrap();

        let augmentation = Production::from_str("<E'> ::= <E>").unwrap();

        let builder = SLRTableBuilder::new(&grammar, &augmentation);
        let slr = builder.build();
        insta::assert_display_snapshot!(slr);
    }
}
