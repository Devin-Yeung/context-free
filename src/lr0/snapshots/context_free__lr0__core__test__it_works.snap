---
source: src/lr0/core.rs
expression: closure
---
+----+-------------------------------------------------------------------------------------------------------------------------------------+
|    | 0                                                                                                                                   |
+----+-------------------------------------------------------------------------------------------------------------------------------------+
| 0  | [<E'> -> • <E>, <E> -> • "+" <E> <T>, <E> -> • <T>, <F> -> • "(" ")" <E>, <F> -> • "id", <T> -> • "*" <F> <T>, <T> -> • <F>]        |
+----+-------------------------------------------------------------------------------------------------------------------------------------+
| 1  | [<E> -> • "+" <E> <T>, <E> -> • <T>, <F> -> "(" • ")" <E>, <F> -> • "(" ")" <E>, <F> -> • "id", <T> -> • "*" <F> <T>, <T> -> • <F>] |
+----+-------------------------------------------------------------------------------------------------------------------------------------+
| 2  | [<F> -> "id" •]                                                                                                                     |
+----+-------------------------------------------------------------------------------------------------------------------------------------+
| 3  | [<E'> -> <E> •, <E> -> "+" • <E> <T>]                                                                                               |
+----+-------------------------------------------------------------------------------------------------------------------------------------+
| 4  | [<T> -> <F> •]                                                                                                                      |
+----+-------------------------------------------------------------------------------------------------------------------------------------+
| 5  | [<E> -> <T> •, <T> -> "*" • <F> <T>]                                                                                                |
+----+-------------------------------------------------------------------------------------------------------------------------------------+
| 6  | [<E> -> "+" • <E> <T>, <F> -> "(" ")" • <E>]                                                                                        |
+----+-------------------------------------------------------------------------------------------------------------------------------------+
| 7  | [<E> -> "+" <E> • <T>, <F> -> • "(" ")" <E>, <F> -> • "id", <T> -> • "*" <F> <T>, <T> -> • <F>]                                     |
+----+-------------------------------------------------------------------------------------------------------------------------------------+
| 8  | [<F> -> • "(" ")" <E>, <F> -> • "id", <T> -> "*" <F> • <T>]                                                                         |
+----+-------------------------------------------------------------------------------------------------------------------------------------+
| 9  | [<F> -> "(" ")" <E> •]                                                                                                              |
+----+-------------------------------------------------------------------------------------------------------------------------------------+
| 10 | [<E> -> "+" <E> <T> •, <T> -> "*" • <F> <T>]                                                                                        |
+----+-------------------------------------------------------------------------------------------------------------------------------------+
| 11 | [<T> -> "*" <F> <T> •]                                                                                                              |
+----+-------------------------------------------------------------------------------------------------------------------------------------+
